src/newrelic/collector/client.go

```diff
@@ -316,8 +335,78 @@ func parseProxy(proxy string) (*url.URL, error) {
 	return url, err
 }
 
+// Has all usage stats needed for the spec
+type DataUsage struct {
+	Attempts int
+	PayloadSize int
+	ResponseSize int
+}
+
+// A map of {endpoint: dataUsage} to keep track of usage for all endpoints of a single harvest
+type dataUsageMap struct {
+	wg *sync.WaitGroup
+	// Needs to use a Lock instead of sync.map because use case is load-modify-store
+	dataMap map[string]DataUsage
+	dataMapLock *sync.Mutex
+}
+
+func NewDataUsageMap() dataUsageMap {
+	data := dataUsageMap{}
+	data.wg = new(sync.WaitGroup)
+	data.dataMapLock = new(sync.Mutex)
+	data.dataMap = make(map[string]DataUsage)
+	return data
+}
+
 type clientImpl struct {
 	httpClient *http.Client
+	// sync map allows simple atomic operations on the map so we don't need a lock
+	harvestDataUsageMaps sync.Map
+}
+
+// Waits until all outstanding http requests currently running for this harvest
+// have collected their data usage
+func (c *clientImpl) WaitOnDataUsage(id string) {
+	v, ok := c.harvestDataUsageMaps.Load(id)
+	if ok {
+		t := v.(dataUsageMap)
+		t.wg.Wait()
+	}
+}
+
+func (c *clientImpl) CollectDataUsage(id string) map[string]DataUsage {
+	v, ok := c.harvestDataUsageMaps.LoadAndDelete(id)
```

Comment:
LoadAndDelete() was introduced in go 1.15 which is beyond our current minimum go version